File: ios/Runner/Whisper/WhisperFlutterBridge.swift
//
//  WhisperFlutterBridge.swift
//  Runner
//
//  Created by Bruce Burgess on 7/16/25.
//

import Foundation
import Flutter
import WhisperCore

fileprivate enum WhisperMethods: String {
    case permission = "callRequestRecordPermission"
    case initialzeModel = "initializeModel"
    case startRecording = "startRecording"
    case stopRecording = "stopRecording"
    case toggleRecording = "toggleRecording"
    case transcribeSample = "transcribeSample"
    case enablePlayback = "enablePlayback"
    case reset = "reset"
    case canTranscribe = "canTranscribe"
    case isRecording = "isRecording"
    case isModelLoaded = "isModelLoaded"
    case getMessageLogs = "getMessageLogs"
}

@MainActor
class WhisperFlutterBridge: NSObject, WhisperDelegate {
    static let shared = WhisperFlutterBridge()
    
    private let whisper = Whisper()
    private var eventSink: FlutterEventSink?
    
    override init() {
        super.init()
        whisper.delegate = self
    }
    
    func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let method  = WhisperMethods(rawValue: call.method) else {
            result(FlutterMethodNotImplemented)
            return
        }
        
        switch method {
        case .permission:
              whisper.callRequestRecordPermission()
              result(nil)
              
          case .initialzeModel:
              guard let args = call.arguments as? [String: Any],
                    let path = args["path"] as? String else {
                  result(FlutterError(code: "BAD_ARGS", message: "Missing model path", details: nil))
                  return
              }

            print("Checking if model exists at path: \(path)")
            guard FileManager.default.fileExists(atPath: path) else {
                result(FlutterError(code: "FILE_NOT_FOUND", message: "Model file does not exist at path: \(path)", details: nil))
                return
            }
                
              Task {
                  do {
                      try await whisper.initializeModel(at: path)
                      result(true)
                  } catch {
                      result(FlutterError(code: "INIT_FAIL", message: error.localizedDescription, details: nil))
                  }
              }

          case .startRecording:
              Task {
                  await whisper.startRecording()
                  result(nil)
              }

          case .stopRecording:
              Task {
                  await whisper.stopRecording()
                  result(nil)
              }

          case .toggleRecording:
              Task {
                  await whisper.toggleRecording()
                  result(nil)
              }

          case .transcribeSample:
              guard let args = call.arguments as? [String: Any],
                    let path = args["path"] as? String else {
                  result(FlutterError(code: "BAD_ARGS", message: "Missing sample path", details: nil))
                  return
              }
              let url = URL(fileURLWithPath: path)
              Task {
                  await whisper.transcribeSample(from: url)
                  result(nil)
              }

          case .enablePlayback:
              guard let args = call.arguments as? [String: Any],
                    let enabled = args["enabled"] as? Bool else {
                  result(FlutterError(code: "BAD_ARGS", message: "Missing enabled flag", details: nil))
                  return
              }
              whisper.enablePlayback(enabled)
              result(nil)

          case .reset:
              whisper.reset()
              result(nil)

          case .canTranscribe:
              result(whisper.canTranscribe())

          case .isRecording:
              result(whisper.isRecording())

          case .isModelLoaded:
              result(whisper.isModelLoaded())

          case .getMessageLogs:
              result(whisper.getMessageLogs())
          }
    }
    
    // Called when transcription is complete
    nonisolated func didTranscribe(_ text: String) {
        Task { @MainActor in
            eventSink?(["event": "didTranscribe", "text": text])
        }
        
    }

    // Called when recording fails
    nonisolated func recordingFailed(_ error: Error) {
        Task { @MainActor in
            eventSink?(["event": "recordingFailed", "error": error.localizedDescription])
        }
    }

    // Called when transcription fails
    nonisolated func failedToTranscribe(_ error: Error) {
        Task { @MainActor in
            eventSink?(["event": "failedToTranscribe", "error": error.localizedDescription])
        }
    }

}

extension WhisperFlutterBridge: FlutterStreamHandler {
    nonisolated func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        Task { @MainActor in
            self.eventSink = events
        }
        return nil
    }

    nonisolated func onCancel(withArguments arguments: Any?) -> FlutterError? {
        Task { @MainActor in
            self.eventSink = nil
        }
        return nil
    }
}

File: android/app/src/main/kotlin/com/redravencomputing/stt_llm_demo/WhisperFlutterBridge.kt
package com.redravencomputing.stt_llm_demo

import android.content.Context
import android.util.Log
import com.redravencomputing.whispercore.Whisper
import com.redravencomputing.whispercore.WhisperDelegate
import com.redravencomputing.whispercore.WhisperLoadError
import com.redravencomputing.whispercore.WhisperOperationError
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File

private enum class WhisperMethods(val methodName: String) {
	Permission("callRequestRecordPermission"),
	InitializeModel("initializeModel"),
	StartRecording("startRecording"),
	StopRecording("stopRecording"),
	ToggleRecording("toggleRecording"),
	TranscribeSample("transcribeSample"),
	EnablePlayback("enablePlayback"),
	Reset("reset"),
	CanTranscribe("canTranscribe"),
	IsRecording("isRecording"),
	IsModelLoaded("isModelLoaded"),
	GetMessageLogs("getMessageLogs"),

	IsMicrophonePermissionGranted("isMicrophonePermissionGranted");

	companion object {
		fun fromMethodName(name: String): WhisperMethods? {
			return entries.find { it.methodName == name }
		}
	}
}

class WhisperFlutterBridge(context: Context) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

	private var eventSink: EventChannel.EventSink? = null

	private fun serializeError(error: WhisperOperationError): String {
		return when (error) {
			is WhisperOperationError.MicPermissionDenied -> "Microphone access denied"
			is WhisperOperationError.MissingRecordedFile -> "Missing Recorded File"
			is WhisperOperationError.ModelNotLoaded -> "Model Not Loaded"
			is WhisperOperationError.RecordingFailed -> "Recording Failed"
			else -> "Unknown error: ${error.message}"
		}
	}

	private val whisper = Whisper(context).apply {
		delegate = object : WhisperDelegate {
			override fun didTranscribe(text: String) {
				eventSink?.success(mapOf("event" to "didTranscribe", "text" to text))
			}

			override fun recordingFailed(error: WhisperOperationError) {
				eventSink?.success(
					mapOf(
						"event" to "recordingFailed",
						"error" to serializeError(error)
					)
				)
			}

			override fun failedToTranscribe(error: WhisperOperationError) {
				eventSink?.success(
					mapOf(
						"event" to "failedToTranscribe",
						"error" to serializeError(error)
					)
				)
			}

			override fun permissionRequestNeeded() {
				eventSink?.success(mapOf("event" to "permissionRequestNeeded"))
			}

			override fun didStartRecording() {
				Log.d("WhisperFlutterBridge", "Did Start Recording")
				eventSink?.success(mapOf("event" to "didStartRecording", "isRecording" to true))
			}

			override fun didStopRecording() {
				Log.d("WhisperFlutterBridge", "Did Stop Recording")
				eventSink?.success(mapOf("event" to "didStopRecording", "isRecording" to false))
			}
		}
	}

	override fun onMethodCall(
		call: MethodCall,
		result: MethodChannel.Result
	) {
		val method = WhisperMethods.fromMethodName(call.method)
		if (method == null) {
			result.notImplemented()
			return
		}
		when(method) {
			WhisperMethods.Permission -> {
				Log.d("WhisperFlutterBridge", "Permission is being called")
				whisper.callRequestRecordPermission()
				result.success(null)
			}
			WhisperMethods.InitializeModel -> {
				val path = call.argument<String>("path")
				if (path == null) {
					result.error("BAD_ARGS", "Missing model path", null)
					return
				}
				CoroutineScope(Dispatchers.Main).launch {
					try {
						whisper.initializeModel(path, false)
						result.success(true)
					} catch (e: WhisperLoadError) {
						result.error("INIT_FAIL", e.localizedMessage, null)
					} catch (e: Exception) {
						result.error("INIT_FAIL", e.localizedMessage, null)
					}
				}
			}
			WhisperMethods.StartRecording -> {
				whisper.startRecording()
				result.success(null)
			}
			WhisperMethods.StopRecording -> {
				whisper.stopRecording()
				result.success(null)
			}
			WhisperMethods.ToggleRecording -> {
				whisper.toggleRecording()
				result.success(null)
			}
			WhisperMethods.TranscribeSample -> {
				val path = call.argument<String>("path")
				if (path == null) {
					result.error("BAD_ARGS", "Missing sample path", null)
					return
				}
				whisper.transcribeAudioFile(
					File(path))
				result.success(null)
			}
			WhisperMethods.EnablePlayback -> {
				val enabled = call.argument<Boolean>("enabled") ?: false
				whisper.enablePlayback(enabled)
				result.success(null)
			}
			WhisperMethods.Reset -> {
				whisper.reset()
				result.success(null)
			}
			WhisperMethods.CanTranscribe ->  result.success(whisper.canTranscribe)
			WhisperMethods.IsRecording ->  result.success(whisper.isRecording)
			WhisperMethods.IsModelLoaded ->  result.success(whisper.isModelLoaded)
			WhisperMethods.GetMessageLogs ->  result.success(whisper.getMessageLogs())
			WhisperMethods.IsMicrophonePermissionGranted -> result.success(whisper.isMicrophonePermissionGranted())
		}
	}

	override fun onListen(
		arguments: Any?,
		events: EventChannel.EventSink?
	) {
		eventSink = events
	}

	override fun onCancel(arguments: Any?) {
		eventSink = null
	}

	fun onRecordPermissionResult(granted: Boolean) {
		whisper.onRecordPermissionResult(granted)
	}

}